fix: handle ingress link correctly when path contains regex

akram opened the pull request:
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Closes #12254

jaideepr97 commented:
Hey @akram 
Could you take a look at the failing unit test?

akram commented:
@jaideepr97 just done. It seems to be passing now.

Marvin9 commented:
Maybe trim the brackets too? eg. /foo/bar/[A-Z0-9]{3}

jaideepr97 commented:
There seems to be an e2e test failure now :sweat_smile:

akram commented:
> Maybe trim the brackets too? eg. /foo/bar/[A-Z0-9]{3}

You mean that everything after an opening bracket can also be remove, right?

Marvin9 commented:
> You mean that everything after an opening bracket can also be remove, right?

Yes

akram commented:
@jaideepr97 and @Marvin9 , I have added the case of brackets, it was less trivial than expected (if you see a better way to get the `IntMin` function done, please review).

Before adding that, I rebased and e2e tests were passing.

akram commented:
@ohauer @jessesuen can you PTAL ?

jannfis commented:
I'm not sure whether it would lead down into a rabbit hole to try and strip any trailing path segments that might be regexp. 

I was wondering, when we detect a `path` in the Ingress that is a regexp, if we should just set it to the empty string so that no link will be displayed. I have the feeling that it's impossible to come up with a proper link for these cases because what we would end up with is most likely an unusable link, anyway.

akram commented:
This fix will not be fixing all possible cases that's true. Especially, given the fact that regex based path for `Ingress` is implementation specific.
But for many scenarios, it should provide decent results.

jannfis commented:
>  But for many scenarios, it should provide decent results.

Given the following path:

`/some/path/(either|here|or|there)/somewhere/else`

I wonder if

`/some/path`

would be a decent result, or would it result in the same 404 or 403 (or even 500) in most cases?

akram commented:
and generating a random string matching the provided regex? Is that something that you consider being a rabbit hole ? :-)

we can use: https://lucasjones.github.io/reggen/
